spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,333 @@
+
+In order to ensure convergence even in the presence of failures, the network must be connected in such a way to allow communication around any malicious nodes and to provide redundant paths between correct ones.
+This may not be feasible at all times, but should happen at least during periods in which the system is stable.
+
+In other words, during periods in which messages are timely delivered between correct processes long enough for multiple communication rounds to succeed, processes will identify and solve differences in their states.
+We call "long enough" $\Delta$.
+
+| Eventual $\Delta$-Timely Convergence |
+|---|
+| If $e\in t_p$, for some correct process $p$, at instant $t$, then by $\text{max} \{t,\text{GST}\} + \Delta$, either $e \in t_q$, for every correct process $q$ or $\bar{e} \in t_p$.
+
+$\Delta$ encapsulates the assumption that, during stable periods, timeouts eventually do not expire precociously, given that they all can be adjusted to reasonable values, and the steps needed to converge on entries can be accomplished within $\Delta$.
+Without precocious timeouts, all new entries should help algorithms progress.
+In the Tendermint algorithm, for example, no votes for Nil are added and the round should end if a proposal was made.
+
+
+
+## The tuple space as a CRDT
+
+The two approaches described in the [previous section](#nodes-state-as-a-tuple-space), without the deletion of entries, correspond to operation-based and state-based [Grow-only SET](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#G-Set_(Grow-only_Set)) CRDT (G-Set).
+This distributed data-structure is easily described as a set per process in which elements are added to include them G-Set; the sets kept by processes are approximations of the G-Set.
+
+The [2 Phase-Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#2P-Set_(Two-Phase_Set)) (2P-Set) is a variation that allows removals.
+It combines two sets, one in which elements are added to include them in the 2P-Set, $A$, and one in which elements are added to remove them from the 2P-Set, $D$; the actual membership of the 2P-Set is given by $A \setminus D$.
+
+
+
+
+> Warning/TODO: A word about tombstones, that is $D$
+> - Only state that is not required should be deleted/tombstone'd.
+> - Instead of tombstones, add new entries that trigger removal of other entries (for example, state about a new height); each node must be given information to realize by itself that an entry is no longer needed.
+> - Tombstones are an optimization, kept to prevent data recreation and redeletion.
+> - Tombstones should be garbage collected at some point; imprecision shouldn't affect correctness/termination, since this is an optimization (as long as deleted state is never required again).
+> - Tombstones are not to be gossiped; if they were, they would need to carry proof for the reason they were created, defeating their point.
```

Comment:
In a 2P-Set, both the add set `A` and the remove set `D` are grow-only sets, i.e. G-Sets: they only accrue elements over time, they never lose elements. Elements in the remove set `D` can be considered tombstones, but those tombstones are not optimizations: they're equally important to elements in the add set `A`, when determining the ultimate set membership. 

It's possible to GC elements in a 2P-Set, but that GC can't be done on the "tombstones" in the remove set in isolation. It has to be performed holistically, to "reset" the state of both the add and remove sets together, and in a way that is reliably the same across all nodes in the network.

In summary, this description of a 2P-Set — and specifically the description of the remove set subcomponent of a 2P-Set — has some problems, at least as far as I understand the concepts at play. I'm happy to be corrected! But based on my current understanding, I don't think the state space can be modeled as a CRDT, as described here..

Reply:
Thanks for the comments, @peterbourgon. The quoted part was a set of notes for myself, things to think about, so please disregard them as they may indeed be imprecise. But just to share what I was thinking, the idea here is that the tombstone can be used to GC elements from A and prevent them from being re-added and if tombstones are gossiped, then the removals occur everywhere, but gossiping the tombstones is not ideal either. So we are working on a way to have implicit tombstones, but the idea is still being developed.

Reply:
Got it.

Why is gossiping tombstones not ideal?

Reply:
Say a message A is in the set and a message B arrives, causing A to become useless. We GC A and create a tombstone nA (not A). If nA is gossiped to another node, then it must carry B as well, to prove that nA was created in good faith, but in this case the receiver of nA can simply apply the same rule as the sender and use B to remove nA. Gossiping nA does not contribute in any way here.

Reply:
Ah, OK, and that all rests on the assumption that Gossip Communication provides reliable delivery of messages to nodes, correct?